In [2]:
!pip install duckdb

In [3]:
import duckdb
import pandas as pd

matches = pd.read_csv("C:/Users/anujm/OneDrive/Documents/Desktop/CricketIQ/data/raw/matches.csv")
deliveries = pd.read_csv("C:/Users/anujm/OneDrive/Documents/Desktop/CricketIQ/data/raw/deliveries.csv")

In [4]:
matches.head()

,id,season,city,date,match_type,player_of_match,venue,team1,team2,toss_winner,toss_decision,winner,result,result_margin,target_runs,target_overs,super_over,method,umpire1,umpire2
0,335982,2007/08,Bangalore,2008-04-18,League,BB McCullum,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140.0,223.0,20.0,N,NaN,Asad Rauf,RE Koertzen
1,335983,2007/08,Chandigarh,2008-04-19,League,MEK Hussey,"Punjab Cricket Association Stadium, Mohali",Kings XI Punjab,Chennai Super Kings,Chennai Super Kings,bat,Chennai Super Kings,runs,33.0,241.0,20.0,N,NaN,MR Benson,SL Shastri
2,335984,2007/08,Delhi,2008-04-19,League,MF Maharoof,Feroz Shah Kotla,Delhi Daredevils,Rajasthan Royals,Rajasthan Royals,bat,Delhi Daredevils,wickets,9.0,130.0,20.0,N,NaN,Aleem Dar,GA Pratapkumar
3,335985,2007/08,Mumbai,2008-04-20,League,MV Boucher,Wankhede Stadium,Mumbai Indians,Royal Challengers Bangalore,Mumbai Indians,bat,Royal Challengers Bangalore,wickets,5.0,166.0,20.0,N,NaN,SJ Davis,DJ Harper
4,335986,2007/08,Kolkata,2008-04-20,League,DJ Hussey,Eden Gardens,Kolkata Knight Riders,Deccan Chargers,Deccan Chargers,bat,Kolkata Knight Riders,wickets,5.0,111.0,20.0,N,NaN,BF Bowden,K Hariharan


In [5]:
deliveries.head()

,match_id,inning,batting_team,bowling_team,over,ball,batter,bowler,non_striker,batsman_runs,extra_runs,total_runs,extras_type,is_wicket,player_dismissed,dismissal_kind,fielder
0,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,1,SC Ganguly,P Kumar,BB McCullum,0,1,1,legbyes,0,NaN,NaN,NaN
1,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,2,BB McCullum,P Kumar,SC Ganguly,0,0,0,NaN,0,NaN,NaN,NaN
2,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,BB McCullum,P Kumar,SC Ganguly,0,1,1,wides,0,NaN,NaN,NaN
3,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,4,BB McCullum,P Kumar,SC Ganguly,0,0,0,NaN,0,NaN,NaN,NaN
4,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,5,BB McCullum,P Kumar,SC Ganguly,0,0,0,NaN,0,NaN,NaN,NaN


In [6]:
matches.shape, deliveries.shape

((1095, 20), (260920, 17))

In [7]:
con = duckdb.connect()

con.register("matches", matches)
con.register("deliveries", deliveries)

In [8]:
con.execute("""
SELECT *
FROM matches
LIMIT 5
""").df()

,id,season,city,date,match_type,player_of_match,venue,team1,team2,toss_winner,toss_decision,winner,result,result_margin,target_runs,target_overs,super_over,method,umpire1,umpire2
0,335982,2007/08,Bangalore,2008-04-18,League,BB McCullum,M Chinnaswamy Stadium,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,Kolkata Knight Riders,runs,140.0,223.0,20.0,N,None,Asad Rauf,RE Koertzen
1,335983,2007/08,Chandigarh,2008-04-19,League,MEK Hussey,"Punjab Cricket Association Stadium, Mohali",Kings XI Punjab,Chennai Super Kings,Chennai Super Kings,bat,Chennai Super Kings,runs,33.0,241.0,20.0,N,None,MR Benson,SL Shastri
2,335984,2007/08,Delhi,2008-04-19,League,MF Maharoof,Feroz Shah Kotla,Delhi Daredevils,Rajasthan Royals,Rajasthan Royals,bat,Delhi Daredevils,wickets,9.0,130.0,20.0,N,None,Aleem Dar,GA Pratapkumar
3,335985,2007/08,Mumbai,2008-04-20,League,MV Boucher,Wankhede Stadium,Mumbai Indians,Royal Challengers Bangalore,Mumbai Indians,bat,Royal Challengers Bangalore,wickets,5.0,166.0,20.0,N,None,SJ Davis,DJ Harper
4,335986,2007/08,Kolkata,2008-04-20,League,DJ Hussey,Eden Gardens,Kolkata Knight Riders,Deccan Chargers,Deccan Chargers,bat,Kolkata Knight Riders,wickets,5.0,111.0,20.0,N,None,BF Bowden,K Hariharan


In [9]:
con.execute("""
SELECT *
FROM deliveries
LIMIT 5
""").df()

,match_id,inning,batting_team,bowling_team,over,ball,batter,bowler,non_striker,batsman_runs,extra_runs,total_runs,extras_type,is_wicket,player_dismissed,dismissal_kind,fielder
0,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,1,SC Ganguly,P Kumar,BB McCullum,0,1,1,legbyes,0,None,None,None
1,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,2,BB McCullum,P Kumar,SC Ganguly,0,0,0,None,0,None,None,None
2,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,BB McCullum,P Kumar,SC Ganguly,0,1,1,wides,0,None,None,None
3,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,4,BB McCullum,P Kumar,SC Ganguly,0,0,0,None,0,None,None,None
4,335982,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,5,BB McCullum,P Kumar,SC Ganguly,0,0,0,None,0,None,None,None


# CricketIQ SQL Analysis 

## Module 0: Data Cleaning and Preparation 

In [10]:
# Create cleaned DuckDB tables in one step

con.execute("""
CREATE OR REPLACE TABLE matches_clean AS

SELECT
    * EXCLUDE (
        season,
        date,
        team1,
        team2,
        toss_winner,
        winner,
        city,
        venue
    ),

    /* Standardize season values */
    CASE
        WHEN CAST(season AS VARCHAR) = '2007/08' THEN '2008'
        WHEN CAST(season AS VARCHAR) = '2009/10' THEN '2010'
        WHEN CAST(season AS VARCHAR) = '2020/21' THEN '2020'
        ELSE CAST(season AS VARCHAR)
    END AS season,

    /* Convert date to proper DATE datatype */
    CAST(date AS DATE) AS date,

    /* Standardize Team 1 */
    CASE
        WHEN team1 = 'Delhi Daredevils'
            THEN 'Delhi Capitals'
        WHEN team1 = 'Kings XI Punjab'
            THEN 'Punjab Kings'
        WHEN team1 = 'Royal Challengers Bangalore'
            THEN 'Royal Challengers Bengaluru'
        WHEN team1 = 'Rising Pune Supergiant'
            THEN 'Rising Pune Supergiants'
        ELSE team1
    END AS team1,

    /* Standardize Team 2 */
    CASE
        WHEN team2 = 'Delhi Daredevils'
            THEN 'Delhi Capitals'
        WHEN team2 = 'Kings XI Punjab'
            THEN 'Punjab Kings'
        WHEN team2 = 'Royal Challengers Bangalore'
            THEN 'Royal Challengers Bengaluru'
        WHEN team2 = 'Rising Pune Supergiant'
            THEN 'Rising Pune Supergiants'
        ELSE team2
    END AS team2,

    /* Standardize toss winner */
    CASE
        WHEN toss_winner = 'Delhi Daredevils'
            THEN 'Delhi Capitals'
        WHEN toss_winner = 'Kings XI Punjab'
            THEN 'Punjab Kings'
        WHEN toss_winner = 'Royal Challengers Bangalore'
            THEN 'Royal Challengers Bengaluru'
        WHEN toss_winner = 'Rising Pune Supergiant'
            THEN 'Rising Pune Supergiants'
        ELSE toss_winner
    END AS toss_winner,

    /* Standardize match winner */
    CASE
        WHEN winner = 'Delhi Daredevils'
            THEN 'Delhi Capitals'
        WHEN winner = 'Kings XI Punjab'
            THEN 'Punjab Kings'
        WHEN winner = 'Royal Challengers Bangalore'
            THEN 'Royal Challengers Bengaluru'
        WHEN winner = 'Rising Pune Supergiant'
            THEN 'Rising Pune Supergiants'
        ELSE winner
    END AS winner,

    /* Standardize city names */
    CASE
        WHEN city = 'Bangalore'
            THEN 'Bengaluru'
        ELSE city
    END AS city,

    /* Standardize venue names */
    CASE
        WHEN venue = 'Subrata Roy Sahara Stadium'
            THEN 'Maharashtra Cricket Association Stadium, Pune'

        WHEN venue = 'Feroz Shah Kotla'
            THEN 'Arun Jaitley Stadium, Delhi'

        WHEN venue = 'M Chinnaswamy Stadium'
            THEN 'M Chinnaswamy Stadium, Bengaluru'

        WHEN venue = 'Rajiv Gandhi International Stadium, Uppal'
            THEN 'Rajiv Gandhi International Stadium, Uppal, Hyderabad'

        WHEN venue = 'Eden Gardens, Kolkata'
            THEN 'Eden Gardens'

        WHEN venue = 'Punjab Cricket Association Stadium, Mohali'
            THEN 'Punjab Cricket Association IS Bindra Stadium'

        WHEN venue = 'Punjab Cricket Association IS Bindra Stadium, Mohali, Chandigarh'
            THEN 'Punjab Cricket Association IS Bindra Stadium'

        WHEN venue = 'Brabourne Stadium, Mumbai'
            THEN 'Brabourne Stadium'

        WHEN venue = 'Wankhede Stadium, Mumbai'
            THEN 'Wankhede Stadium'
        WHEN venue = 'MA Chidambaram Stadium, Chepauk'
             THEN 'MA Chidambaram Stadium'

        ELSE venue
    END AS venue

FROM matches;
""")

con.execute("""
CREATE OR REPLACE TABLE deliveries_clean AS

SELECT
    * EXCLUDE (
        batting_team,
        bowling_team
    ),

    /* Standardize batting team */
    CASE
        WHEN batting_team = 'Delhi Daredevils'
            THEN 'Delhi Capitals'
        WHEN batting_team = 'Kings XI Punjab'
            THEN 'Punjab Kings'
        WHEN batting_team = 'Royal Challengers Bangalore'
            THEN 'Royal Challengers Bengaluru'
        WHEN batting_team = 'Rising Pune Supergiant'
            THEN 'Rising Pune Supergiants'
        ELSE batting_team
    END AS batting_team,

    /* Standardize bowling team */
    CASE
        WHEN bowling_team = 'Delhi Daredevils'
            THEN 'Delhi Capitals'
        WHEN bowling_team = 'Kings XI Punjab'
            THEN 'Punjab Kings'
        WHEN bowling_team = 'Royal Challengers Bangalore'
            THEN 'Royal Challengers Bengaluru'
        WHEN bowling_team = 'Rising Pune Supergiant'
            THEN 'Rising Pune Supergiants'
        ELSE bowling_team
    END AS bowling_team

FROM deliveries;
""")

print("Data cleaning completed successfully.")

Data cleaning completed successfully.


In [11]:
con.execute("""
SELECT
    (SELECT COUNT(*) FROM matches_clean) AS matches_rows,
    (SELECT COUNT(*) FROM deliveries_clean) AS deliveries_rows
""").df()

,matches_rows,deliveries_rows
0,1095,260920


In [12]:
con.execute("""
SELECT DISTINCT season
FROM matches_clean
ORDER BY CAST(season AS INTEGER)
""").df()

,season
0,2008
1,2009
2,2010
3,2011
4,2012
5,2013
6,2014
7,2015
8,2016
9,2017


In [13]:
con.execute("""
SELECT DISTINCT team1 AS team
FROM matches_clean
ORDER BY team
""").df()

,team
0,Chennai Super Kings
1,Deccan Chargers
2,Delhi Capitals
3,Gujarat Lions
4,Gujarat Titans
5,Kochi Tuskers Kerala
6,Kolkata Knight Riders
7,Lucknow Super Giants
8,Mumbai Indians
9,Pune Warriors


## Module 1: Basic SQL and Match Analytics 

### Query 1: Total IPL Matches and Seasons

#### Business Problem

Determine the total number of IPL matches and the number of seasons covered in the dataset.

In [14]:
query_1 = """
SELECT
    COUNT(*) AS total_matches,
    COUNT(DISTINCT season) AS total_seasons
FROM matches_clean
"""

con.execute(query_1).df()

,total_matches,total_seasons
0,1095,17


### Observation

The dataset contains **1095 IPL matches** across **17 unique seasons**, providing a substantial historical record for analysing IPL trends and performances.

### Query 2: Number of Matches Played in Each Season

#### Business Problem

Analyse how the number of IPL matches has changed across different seasons.

In [15]:
query_2 = """
SELECT
    season,
    COUNT(*) AS total_matches
FROM matches_clean
GROUP BY season
ORDER BY season
"""

con.execute(query_2).df()

,season,total_matches
0,2008,58
1,2009,57
2,2010,60
3,2011,73
4,2012,74
5,2013,76
6,2014,60
7,2015,59
8,2016,60
9,2017,59


### Observation

The number of matches varies across IPL seasons due to changes in the number of participating teams and tournament format. Seasons with more franchises generally contain a higher number of matches.

### Query 3: Cities Hosting the Most IPL Matches

#### Business Problem

Identify the cities that have hosted the highest number of IPL matches.

In [16]:
query_3 = """
SELECT
    city,
    COUNT(*) AS matches_hosted
FROM matches_clean
WHERE city IS NOT NULL
GROUP BY city
ORDER BY matches_hosted DESC
LIMIT 10
"""

con.execute(query_3).df()

,city,matches_hosted
0,Mumbai,173
1,Bengaluru,94
2,Kolkata,93
3,Delhi,90
4,Chennai,85
5,Hyderabad,77
6,Chandigarh,61
7,Jaipur,57
8,Pune,51
9,Abu Dhabi,37


### Observation

Major IPL centres such as Mumbai, Bengaluru, Kolkata, Delhi and Chennai dominate the list because they host established franchises and frequently stage home matches.The prescence of Abu Dhabi reflects the IPL seasons that were shifted to the UAE.

### Query 4: Most Frequently Used IPL Venues

#### Business Problem

Determine which stadiums have hosted the greatest number of IPL matches.

In [17]:
query_4 = """
SELECT
    venue,
    COUNT(*) AS matches_hosted
FROM matches_clean
WHERE venue IS NOT NULL
GROUP BY venue
ORDER BY matches_hosted DESC
LIMIT 10
"""

con.execute(query_4).df()

,venue,matches_hosted
0,Wankhede Stadium,118
1,Eden Gardens,93
2,"M Chinnaswamy Stadium, Bengaluru",79
3,"Arun Jaitley Stadium, Delhi",76
4,"Rajiv Gandhi International Stadium, Uppal, Hyd...",62
5,MA Chidambaram Stadium,57
6,Punjab Cricket Association IS Bindra Stadium,50
7,Sawai Mansingh Stadium,47
8,Dubai International Cricket Stadium,46
9,Sheikh Zayed Stadium,29


### Observation
Wankhede Stadium has hosted the highest number of IPL matches (118), making it the most frequently used venue in IPL history.Eden Gardens (93) and M Chinnaswamy Stadium, Bengaluru (79) rank second and third, highlighting their long-standing importance in the tournament.Traditional venues such as Arun Jaitley Stadium, Delhi, MA Chidambaram Stadium, and Sawai Mansingh Stadium consistently appear among the top venues.The presence of Dubai International Cricket Stadium (46) reflects the IPL seasons that were shifted to the UAE.

### Query 5: Toss Decision Distribution

#### Business Problem

Determine whether IPL captains more frequently choose to bat or field after winning the toss.

In [18]:
query_5 = """
SELECT
    toss_decision,
    COUNT(*) AS total_matches,
    ROUND(
        COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (),
        2
    ) AS percentage
FROM matches_clean
WHERE toss_decision IS NOT NULL
GROUP BY toss_decision
ORDER BY total_matches DESC
"""

con.execute(query_5).df()

,toss_decision,total_matches,percentage
0,field,704,64.29
1,bat,391,35.71


### Observation

Captains have generally preferred to **field** after winning the toss. This may reflect the importance of chasing, dew conditions and greater clarity about the required target.

## Module 2: Team Analytics

### Query 6: Teams with the Most IPL Matches Played

#### Business Problem

Identify the teams that have participated in the highest number of IPL matches.

In [19]:
con.execute("""
WITH team_matches AS (

    SELECT team1 AS team
    FROM matches_clean

    UNION ALL

    SELECT team2 AS team
    FROM matches_clean
)

SELECT
    team,
    COUNT(*) AS matches_played
FROM team_matches
GROUP BY team
ORDER BY matches_played DESC;
""").df()

,team,matches_played
0,Mumbai Indians,261
1,Royal Challengers Bengaluru,255
2,Delhi Capitals,252
3,Kolkata Knight Riders,251
4,Punjab Kings,246
5,Chennai Super Kings,238
6,Rajasthan Royals,221
7,Sunrisers Hyderabad,182
8,Deccan Chargers,75
9,Pune Warriors,46


### Observation 

Mumbai Indians have played the highest number of IPL matches (261), followed by Royal Challengers Bengaluru (255) and Delhi Capitals (252). Teams like Chennai Super Kings and Rajasthan Royals have played lesser matches due to their 2 year ban from the IPL, while newer franchises such as Gujarat Titans and Lucknow Super Giants have naturally played fewer matches due to their recent introduction to the league.

### Query 7: Teams with the Most IPL Match Wins

#### Business Problem

Determine which teams have won the highest number of IPL matches.

In [20]:
con.execute("""
SELECT
    winner AS team,
    COUNT(*) AS matches_won
FROM matches_clean
WHERE winner IS NOT NULL
GROUP BY winner
ORDER BY matches_won DESC, team;
""").df()

,team,matches_won
0,Mumbai Indians,144
1,Chennai Super Kings,138
2,Kolkata Knight Riders,131
3,Royal Challengers Bengaluru,123
4,Delhi Capitals,115
5,Punjab Kings,112
6,Rajasthan Royals,112
7,Sunrisers Hyderabad,88
8,Deccan Chargers,29
9,Gujarat Titans,28


### Observation 

Mumbai Indians lead the league with 144 match victories, followed by Chennai Super Kings (138) and Kolkata Knight Riders (131). Despite playing fewer matches than some franchises, Chennai Super Kings have maintained one of the highest win totals, reflecting their consistent performance across IPL seasons.

### Query 8: Team Win Percentage

#### Business Problem

Compare teams based on their overall match win percentage.

In [21]:
con.execute("""
WITH matches_played AS (

    SELECT team1 AS team
    FROM matches_clean

    UNION ALL

    SELECT team2 AS team
    FROM matches_clean
),

team_matches AS (

    SELECT
        team,
        COUNT(*) AS matches_played
    FROM matches_played
    GROUP BY team
),

team_wins AS (

    SELECT
        winner AS team,
        COUNT(*) AS matches_won
    FROM matches_clean
    WHERE winner IS NOT NULL
    GROUP BY winner
)

SELECT
    tm.team,
    tm.matches_played,
    COALESCE(tw.matches_won, 0) AS matches_won,

    ROUND(
        100.0 * COALESCE(tw.matches_won, 0)
        / tm.matches_played,
        2
    ) AS win_percentage

FROM team_matches tm

LEFT JOIN team_wins tw
    ON tm.team = tw.team

ORDER BY win_percentage DESC;
""").df()

,team,matches_played,matches_won,win_percentage
0,Gujarat Titans,45,28,62.22
1,Chennai Super Kings,238,138,57.98
2,Mumbai Indians,261,144,55.17
3,Lucknow Super Giants,44,24,54.55
4,Kolkata Knight Riders,251,131,52.19
5,Rajasthan Royals,221,112,50.68
6,Rising Pune Supergiants,30,15,50.00
7,Sunrisers Hyderabad,182,88,48.35
8,Royal Challengers Bengaluru,255,123,48.24
9,Delhi Capitals,252,115,45.63


### Observation 

Gujarat Titans have the highest win percentage (62.22%) among all IPL franchises, despite playing only 45 matches. Among long-established teams, Chennai Super Kings lead with a remarkable 57.98% win rate, followed by Mumbai Indians (55.17%) and Kolkata Knight Riders (52.19%). Pune Warriors have the lowest win percentage at 26.09%.

### Query 9: Toss Wins by Team

#### Business Problem 

Identify the teams that have won the toss most frequently.

In [22]:
con.execute("""
SELECT
    toss_winner AS team,
    COUNT(*) AS tosses_won
FROM matches_clean
WHERE toss_winner IS NOT NULL
GROUP BY toss_winner
ORDER BY tosses_won DESC, team;
""").df()

,team,tosses_won
0,Mumbai Indians,143
1,Delhi Capitals,130
2,Chennai Super Kings,122
3,Kolkata Knight Riders,122
4,Royal Challengers Bengaluru,121
5,Rajasthan Royals,120
6,Punjab Kings,109
7,Sunrisers Hyderabad,88
8,Deccan Chargers,43
9,Gujarat Titans,22


### Observation 

Mumbai Indians have won the toss the most times (143), followed by Delhi Capitals (130) and both Chennai Super Kings and Kolkata Knight Riders (122 each). The toss-win counts generally reflect the number of matches played, with long-established franchises naturally appearing at the top of the list.

### Query 10: Toss Win Converted into Match Win

### Business Problem

Measure how frequently each team won the match after winning the toss.

In [23]:
con.execute("""
SELECT
    toss_winner AS team,
    COUNT(*) AS tosses_won,

    SUM(
        CASE
            WHEN toss_winner = winner THEN 1
            ELSE 0
        END
    ) AS matches_won_after_toss,

    ROUND(
        100.0 *
        SUM(
            CASE
                WHEN toss_winner = winner THEN 1
                ELSE 0
            END
        )
        / COUNT(*),
        2
    ) AS toss_conversion_percentage

FROM matches_clean

WHERE toss_winner IS NOT NULL

GROUP BY toss_winner 
ORDER BY toss_conversion_percentage DESC;
""").df()

,team,tosses_won,matches_won_after_toss,toss_conversion_percentage
0,Gujarat Lions,15,10.0,66.67
1,Gujarat Titans,22,14.0,63.64
2,Rising Pune Supergiants,13,8.0,61.54
3,Chennai Super Kings,122,75.0,61.48
4,Kolkata Knight Riders,122,68.0,55.74
5,Mumbai Indians,143,78.0,54.55
6,Lucknow Super Giants,19,10.0,52.63
7,Royal Challengers Bengaluru,121,61.0,50.41
8,Kochi Tuskers Kerala,8,4.0,50.00
9,Rajasthan Royals,120,60.0,50.00


### Observations 

Gujarat Lions recorded the highest toss-to-match conversion rate at 66.67%, followed by Gujarat Titans at 63.64% and Rising Pune Supergiants at 61.54%. Among long-established franchises, Chennai Super Kings performed best, winning 75 of the 122 matches in which they won the toss, for a conversion rate of 61.48%. Pune Warriors had the lowest conversion rate at just 15%.

## Module 3: Batting Analytics 

### Query 11: Top Run Scorers in IPL

#### Business Problem

Identify the batters who have scored the most runs in IPL history.

In [24]:
con.execute("""
SELECT
    batter,
    CAST(SUM(batsman_runs) AS INTEGER) AS total_runs
FROM deliveries_clean
GROUP BY batter
ORDER BY total_runs DESC, batter
LIMIT 10;
""").df()

,batter,total_runs
0,V Kohli,8014
1,S Dhawan,6769
2,RG Sharma,6630
3,DA Warner,6567
4,SK Raina,5536
5,MS Dhoni,5243
6,AB de Villiers,5181
7,CH Gayle,4997
8,RV Uthappa,4954
9,KD Karthik,4843


### Observation 

Virat Kohli is the highest run scorer in IPL history with 8,014 runs, followed by Shikhar Dhawan (6,769) and Rohit Sharma (6,630). These batters have consistently accumulated runs across multiple IPL seasons, highlighting their longevity and reliability in the tournament.

### Query 12: Batters with the Highest Batting Average

#### Business Problem

Identify the batters with the highest batting average, considering only players with a meaningful number of dismissals.

In [25]:
con.execute("""
WITH batter_runs AS (
    SELECT
        batter,
        CAST(SUM(batsman_runs) AS INTEGER) AS total_runs
    FROM deliveries_clean
    GROUP BY batter
),

batter_dismissals AS (
    SELECT
        player_dismissed AS batter,
        COUNT(*) AS dismissals
    FROM deliveries_clean
    WHERE player_dismissed IS NOT NULL
      AND dismissal_kind NOT IN (
          'retired hurt',
          'retired out',
          'obstructing the field'
      )
    GROUP BY player_dismissed
)

SELECT
    br.batter,
    br.total_runs,
    COALESCE(bd.dismissals, 0) AS dismissals,
    ROUND(
        br.total_runs * 1.0 / NULLIF(bd.dismissals, 0),
        2
    ) AS batting_average
FROM batter_runs br
JOIN batter_dismissals bd
    ON br.batter = bd.batter
WHERE bd.dismissals >= 20
ORDER BY batting_average DESC, br.batter
LIMIT 10;
""").df()

,batter,total_runs,dismissals,batting_average
0,B Sai Sudharsan,1034,21,49.24
1,KL Rahul,4689,105,44.66
2,RD Gaikwad,2380,57,41.75
3,DA Warner,6567,164,40.04
4,LMP Simmons,1079,27,39.96
5,Tilak Varma,1156,29,39.86
6,AB de Villiers,5181,130,39.85
7,JP Duminy,2029,51,39.78
8,CH Gayle,4997,126,39.66
9,SE Marsh,2489,63,39.51


### Observation 

B Sai Sudharsan has the highest batting average (49.24) among batters with at least 20 dismissals, followed by KL Rahul (44.66) and Ruturaj Gaikwad (41.75). Established players such as David Warner, AB de Villiers, and Chris Gayle also feature among the top averages, reflecting both consistency and longevity in the IPL.

### Query 13: Batters with the Highest Strike Rate

#### Business Problem

Identify the batters who score at the fastest rate, considering only players who have faced a substantial number of legal deliveries.

In [26]:
con.execute("""
SELECT
    batter,
    CAST(SUM(batsman_runs) AS INTEGER) AS total_runs,

CAST(
    SUM(
        CASE
            WHEN extras_type IS NULL
                 OR extras_type != 'wides'
            THEN 1
            ELSE 0
        END
    ) AS INTEGER 
) AS balls_faced,

    ROUND(
        100.0 * SUM(batsman_runs)
        / NULLIF(
            SUM(
                CASE
                    WHEN extras_type IS NULL
                         OR extras_type != 'wides'
                    THEN 1
                    ELSE 0
                END
            ),
            0
        ),
        2
    ) AS strike_rate

FROM deliveries_clean

GROUP BY batter

HAVING SUM(
    CASE
        WHEN extras_type IS NULL
             OR extras_type != 'wides'
        THEN 1
        ELSE 0
    END
) >= 500

ORDER BY strike_rate DESC, batter
LIMIT 10;
""").df()

,batter,total_runs,balls_faced,strike_rate
0,AD Russell,2488,1423,174.84
1,H Klaasen,993,590,168.31
2,SP Narine,1534,925,165.84
3,LS Livingstone,939,578,162.46
4,N Pooran,1769,1092,162.00
5,RM Patidar,799,503,158.85
6,GJ Maxwell,2772,1769,156.70
7,V Sehwag,2728,1755,155.44
8,Abhishek Sharma,1377,887,155.24
9,SO Hetmyer,1243,812,153.08


### Observation 

Andre Russell has the highest strike rate (174.84) among batters who have faced at least 500 legal deliveries, followed by Heinrich Klaasen (168.31) and Sunil Narine (165.84). The list is dominated by aggressive middle-order batters and power hitters known for scoring rapidly rather than accumulating large totals.

### Query 14: Batters with the Most Fours

In [27]:
con.execute("""
SELECT
    batter,
    CAST(
     SUM(
        CASE
            WHEN batsman_runs = 4 THEN 1
            ELSE 0
        END
        ) AS INTEGER
    ) AS fours
FROM deliveries_clean
GROUP BY batter
ORDER BY fours DESC, batter
LIMIT 10;
""").df()

,batter,fours
0,S Dhawan,768
1,V Kohli,708
2,DA Warner,663
3,RG Sharma,599
4,SK Raina,506
5,G Gambhir,492
6,RV Uthappa,481
7,AM Rahane,479
8,KD Karthik,466
9,F du Plessis,422


### Observation 

Shikhar Dhawan has hit the most fours in IPL history (768), followed by Virat Kohli (708) and David Warner (663). The list is dominated by top-order batters who have consistently accumulated runs through boundaries over multiple IPL seasons.

### Query 15: Batters with the Most Sixes

#### Business Problem

Identify the batters who have hit the highest number of sixes in IPL history.

In [28]:
con.execute("""
SELECT
    batter,
    CAST(
     SUM(
        CASE
            WHEN batsman_runs = 6 THEN 1
            ELSE 0
        END
        ) AS INTEGER 
    ) AS sixes
FROM deliveries_clean
GROUP BY batter
ORDER BY sixes DESC, batter
LIMIT 10;
""").df()

,batter,sixes
0,CH Gayle,359
1,RG Sharma,281
2,V Kohli,273
3,AB de Villiers,253
4,MS Dhoni,252
5,DA Warner,236
6,KA Pollard,224
7,AD Russell,209
8,SV Samson,206
9,SK Raina,204


## Module 4: Bowling Analytics

### Query 16: Top Wicket Takers in IPL

#### Business Problem

Identify the bowlers who have taken the most wickets in IPL history.

In [29]:
con.execute("""
SELECT
    bowler,

    CAST(
        SUM(
            CASE
                WHEN dismissal_kind IN (
                    'bowled',
                    'caught',
                    'caught and bowled',
                    'lbw',
                    'stumped',
                    'hit wicket'
                )
                THEN 1
                ELSE 0
            END
        ) AS INTEGER
    ) AS wickets

FROM deliveries_clean

GROUP BY bowler

ORDER BY wickets DESC, bowler

LIMIT 10;
""").df()

,bowler,wickets
0,YS Chahal,205
1,PP Chawla,192
2,DJ Bravo,183
3,B Kumar,181
4,R Ashwin,180
5,SP Narine,180
6,A Mishra,174
7,SL Malinga,170
8,JJ Bumrah,168
9,RA Jadeja,160


### Observation 

Yuzvendra Chahal is the highest wicket-taker in the dataset with 205 wickets, followed by Piyush Chawla (192) and Dwayne Bravo (183). The list includes both spin and pace bowlers, showing that sustained wicket-taking success in the IPL has come through different bowling styles.

### Query 17: Best Bowling Average

#### Business Problem

Identify the bowlers with the lowest bowling average among those who have taken at least 100 wickets.

In [30]:
con.execute("""
WITH bowling_runs AS (

SELECT
    bowler,
    SUM(total_runs) AS runs_conceded
FROM deliveries_clean
GROUP BY bowler

),

bowling_wickets AS (

SELECT
    bowler,

    COUNT(*) AS wickets

FROM deliveries_clean

WHERE dismissal_kind IN (
    'bowled',
    'caught',
    'caught and bowled',
    'lbw',
    'stumped',
    'hit wicket'
)

GROUP BY bowler

)

SELECT

    br.bowler,

    CAST(br.runs_conceded AS INTEGER) AS runs_conceded,

    bw.wickets,

    ROUND(
        br.runs_conceded * 1.0 / bw.wickets,
        2
    ) AS bowling_average

FROM bowling_runs br

JOIN bowling_wickets bw
ON br.bowler = bw.bowler

WHERE bw.wickets >= 100

ORDER BY bowling_average ASC, br.bowler

LIMIT 10;
""").df()

,bowler,runs_conceded,wickets,bowling_average
0,SL Malinga,3486,170,20.51
1,K Rabada,2652,120,22.10
2,Rashid Khan,3340,149,22.42
3,YS Chahal,4681,205,22.83
4,JJ Bumrah,3840,168,22.86
5,AD Russell,2691,115,23.40
6,HV Patel,3215,135,23.81
7,A Nehra,2537,106,23.93
8,A Mishra,4193,174,24.10
9,DJ Bravo,4436,183,24.24


### Observation 

Lasith Malinga has the best bowling average (20.51) among bowlers with at least 100 wickets, followed by Kagiso Rabada (22.10) and Rashid Khan (22.42). The rankings highlight bowlers who consistently took wickets while conceding relatively few runs over extended IPL careers.

### Query 18: Best Economy Rate

#### Business Problem

Identify the bowlers with the lowest economy rate among those who have bowled at least 300 overs.

In [31]:
con.execute("""
WITH bowling_stats AS (

SELECT

    bowler,

    SUM(total_runs) AS runs_conceded,

    SUM(
        CASE
            WHEN extras_type IS DISTINCT FROM 'wides'
             AND extras_type IS DISTINCT FROM 'noballs'
            THEN 1
            ELSE 0
        END
    ) AS legal_balls

FROM deliveries_clean

GROUP BY bowler

)

SELECT

    bowler,

    CAST(runs_conceded AS INTEGER) AS runs_conceded,

    CAST(legal_balls AS INTEGER) AS balls,

    ROUND(
        runs_conceded * 6.0 / legal_balls,
        2
    ) AS economy

FROM bowling_stats

WHERE legal_balls >= 1800

ORDER BY economy ASC, bowler

LIMIT 10;
""").df()

,bowler,runs_conceded,balls,economy
0,SP Narine,4672,4081,6.87
1,Rashid Khan,3340,2872,6.98
2,DW Steyn,2583,2182,7.10
3,Harbhajan Singh,4101,3416,7.20
4,R Ashwin,5435,4524,7.21
5,AR Patel,3824,3105,7.39
6,SL Malinga,3486,2828,7.40
7,A Mishra,4193,3371,7.46
8,KH Pandya,2644,2122,7.48
9,JJ Bumrah,3840,3075,7.49


### Observation 

Sunil Narine has the best economy rate (6.87 runs per over) among bowlers who have bowled at least 300 overs, followed by Rashid Khan (6.98) and Dale Steyn (7.10). The top rankings are dominated by spinners and disciplined pace bowlers who consistently restricted scoring over long IPL careers.

### Query 19: Best Bowling Strike Rate

#### Business Problem

Identify the bowlers who require the fewest deliveries to take a wicket, considering only bowlers with at least 100 wickets.

In [32]:
con.execute("""
WITH bowling_balls AS (

SELECT

    bowler,

    SUM(
        CASE
            WHEN extras_type IS DISTINCT FROM 'wides'
             AND extras_type IS DISTINCT FROM 'noballs'
            THEN 1
            ELSE 0
        END
    ) AS legal_balls

FROM deliveries_clean

GROUP BY bowler

),

bowling_wickets AS (

SELECT

    bowler,

    COUNT(*) AS wickets

FROM deliveries_clean

WHERE dismissal_kind IN (
    'bowled',
    'caught',
    'caught and bowled',
    'lbw',
    'stumped',
    'hit wicket'
)

GROUP BY bowler

)

SELECT

    bb.bowler,

    CAST(bb.legal_balls AS INTEGER) AS balls,

    bw.wickets,

    ROUND(
        bb.legal_balls * 1.0 / bw.wickets,
        2
    ) AS strike_rate

FROM bowling_balls bb

JOIN bowling_wickets bw
ON bb.bowler = bw.bowler

WHERE bw.wickets >= 100

ORDER BY strike_rate ASC, bb.bowler

LIMIT 10;
""").df()

,bowler,balls,wickets,strike_rate
0,AD Russell,1697,115,14.76
1,K Rabada,1829,120,15.24
2,HV Patel,2162,135,16.01
3,SL Malinga,2828,170,16.64
4,DJ Bravo,3120,183,17.05
5,MM Sharma,2253,132,17.07
6,YS Chahal,3521,205,17.18
7,A Nehra,1908,106,18.00
8,JJ Bumrah,3075,168,18.30
9,Mohammed Shami,2435,127,19.17


### Observation 

Andre Russell has the best bowling strike rate (14.76 balls per wicket) among bowlers with at least 100 wickets, followed by Kagiso Rabada (15.24) and Harshal Patel (16.01). These bowlers consistently take wickets at short intervals, highlighting their effectiveness as attacking wicket-takers in the IPL.

### Query 20: Bowlers with the Most Dot Balls

#### Business Problem

Identify the bowlers who have bowled the most dot balls in IPL history.

In [33]:
con.execute("""
SELECT

    bowler,

    CAST(
        SUM(
            CASE
                WHEN total_runs = 0
                THEN 1
                ELSE 0
            END
        ) AS INTEGER
    ) AS dot_balls

FROM deliveries_clean

GROUP BY bowler

ORDER BY dot_balls DESC, bowler

LIMIT 10;
""").df()

,bowler,dot_balls
0,B Kumar,1632
1,SP Narine,1569
2,R Ashwin,1552
3,PP Chawla,1325
4,Harbhajan Singh,1263
5,JJ Bumrah,1228
6,RA Jadeja,1216
7,YS Chahal,1194
8,UT Yadav,1186
9,A Mishra,1185


### Observation 

Bhuvneshwar Kumar has bowled the most dot balls in IPL history (1,632), followed by Sunil Narine (1,569) and Ravichandran Ashwin (1,552). The top bowlers are predominantly known for their accuracy and ability to consistently restrict scoring opportunities over long IPL careers.

## Module 5: Advanced SQL Analytics

### Query 21: Orange Cap Winner by Season

#### Business Problem

Identify the highest run-scorer in every IPL season using a window-ranking function.

In [34]:
con.execute("""
WITH season_runs AS (

    SELECT
        m.season,
        d.batter,
        SUM(d.batsman_runs) AS total_runs

    FROM deliveries_clean d

    JOIN matches_clean m
        ON d.match_id = m.id

    GROUP BY
        m.season,
        d.batter
),

ranked_batters AS (

    SELECT
        season,
        batter,
        total_runs,

        RANK() OVER (
            PARTITION BY season
            ORDER BY total_runs DESC
        ) AS run_rank

    FROM season_runs
)

SELECT
    season,
    batter,
    CAST(total_runs AS INTEGER) AS runs

FROM ranked_batters

WHERE run_rank = 1

ORDER BY season;
""").df()

,season,batter,runs
0,2008,SE Marsh,616
1,2009,ML Hayden,572
2,2010,SR Tendulkar,618
3,2011,CH Gayle,608
4,2012,CH Gayle,733
5,2013,MEK Hussey,733
6,2014,RV Uthappa,660
7,2015,DA Warner,562
8,2016,V Kohli,973
9,2017,DA Warner,641


### Observation 

Chris Gayle, Virat Kohli and David Warner are the only batters to win the Orange Cap multiple times, with Gayle topping the run charts in 2011 and 2012, Virat Kohli in 2016 and 2024 and Warner in 2015 and 2019. Virat Kohli's 973 runs in 2016 remains the highest run tally by an Orange Cap winner in a single IPL season.

### Query 22: Purple Cap Winner by Season

#### Business Problem

Identify the highest wicket-taker in every IPL season while counting only dismissals officially credited to the bowler.

In [35]:
con.execute("""
WITH season_wickets AS (

    SELECT
        m.season,
        d.bowler,

        SUM(
            CASE
                WHEN d.dismissal_kind IN (
                    'bowled',
                    'caught',
                    'caught and bowled',
                    'lbw',
                    'stumped',
                    'hit wicket'
                )
                THEN 1
                ELSE 0
            END
        ) AS wickets

    FROM deliveries_clean d

    JOIN matches_clean m
        ON d.match_id = m.id

    GROUP BY
        m.season,
        d.bowler
),

ranked_bowlers AS (

    SELECT
        season,
        bowler,
        wickets,

        RANK() OVER (
            PARTITION BY season
            ORDER BY wickets DESC
        ) AS wicket_rank

    FROM season_wickets

    WHERE wickets > 0
)

SELECT
    season,
    bowler,
    CAST(wickets AS INTEGER) AS wickets

FROM ranked_bowlers

WHERE wicket_rank = 1

ORDER BY season;
""").df()

,season,bowler,wickets
0,2008,Sohail Tanvir,22
1,2009,RP Singh,23
2,2010,PP Ojha,21
3,2011,SL Malinga,28
4,2012,M Morkel,25
5,2013,DJ Bravo,32
6,2014,MM Sharma,23
7,2015,DJ Bravo,26
8,2016,B Kumar,23
9,2017,B Kumar,26


### Observation 

Dwayne Bravo, Bhuvneshwar Kumar, Harshal Patel and Kagiso Rabada are the only bowlers to finish as the highest wicket-taker in multiple IPL seasons. The 2019 season produced a tie for the Purple Cap, with Kagiso Rabada and Imran Tahir each taking 26 wickets, demonstrating why a ranking function is more appropriate than selecting only a single maximum.

### Query 23: Top Three Run-Scorers in Every Season

#### Business Problem

Rank batters within each season and identify the top three run-scorers for every IPL season.

In [36]:
con.execute("""
WITH season_runs AS (

    SELECT
        m.season,
        d.batter,
        SUM(d.batsman_runs) AS total_runs

    FROM deliveries_clean d

    JOIN matches_clean m
        ON d.match_id = m.id

    GROUP BY
        m.season,
        d.batter
),

ranked_batters AS (

    SELECT
        season,
        batter,
        total_runs,

        DENSE_RANK() OVER (
            PARTITION BY season
            ORDER BY total_runs DESC
        ) AS season_rank

    FROM season_runs
)

SELECT
    season,
    season_rank,
    batter,
    CAST(total_runs AS INTEGER) AS runs

FROM ranked_batters

WHERE season_rank <= 3

ORDER BY
    season,
    season_rank,
    batter;
""").df()

,season,season_rank,batter,runs
0,2008,1,SE Marsh,616
1,2008,2,G Gambhir,534
2,2008,3,ST Jayasuriya,514
3,2009,1,ML Hayden,572
4,2009,2,AC Gilchrist,495
5,2009,3,AB de Villiers,465
6,2010,1,SR Tendulkar,618
7,2010,2,JH Kallis,572
8,2010,3,SK Raina,528
9,2011,1,CH Gayle,608


### Observation 

The query ranks the top three run-scorers for every IPL season using DENSE_RANK(), preserving tied positions. David Warner appears among the top three in five seasons (2015, 2016, 2017, 2019 and 2020), more than any other batter,Virat Kohli appears amoung the top 3 in four seasons(2011,2013,2016 and 2024) highlighting their remarkable consistency across multiple IPL editions.

### Query 24: Running Team Score by Over

In [46]:
con.execute("""
WITH over_scores AS (

    SELECT
        match_id,
        inning,
        batting_team,
        over,
        SUM(total_runs) AS runs_in_over

    FROM deliveries_clean

    GROUP BY
        match_id,
        inning,
        batting_team,
        over
),

score_progression AS (

    SELECT
        match_id,
        inning,
        batting_team,
        over,
        runs_in_over,

        SUM(runs_in_over) OVER (
            PARTITION BY match_id, inning
            ORDER BY over
            ROWS BETWEEN UNBOUNDED PRECEDING
                     AND CURRENT ROW
        ) AS cumulative_score

    FROM over_scores
)

SELECT
    match_id,
    inning,
    batting_team,
    over + 1 AS over_number,
    CAST(runs_in_over AS INTEGER) AS runs_in_over,
    CAST(cumulative_score AS INTEGER) AS cumulative_score

FROM score_progression

ORDER BY
    match_id,
    inning,
    over

LIMIT 100;
""").df()

,match_id,inning,batting_team,over_number,runs_in_over,cumulative_score
0,335982,1,Kolkata Knight Riders,1,3,3
1,335982,1,Kolkata Knight Riders,2,18,21
2,335982,1,Kolkata Knight Riders,3,6,27
3,335982,1,Kolkata Knight Riders,4,23,50
4,335982,1,Kolkata Knight Riders,5,10,60
...,...,...,...,...,...,...
95,335984,1,Rajasthan Royals,20,8,129
96,335984,2,Delhi Capitals,1,8,8
97,335984,2,Delhi Capitals,2,13,21
98,335984,2,Delhi Capitals,3,5,26


### Observation 

The cumulative score increases over by over by adding each over’s runs to the previous total. For example, Kolkata Knight Riders scored 3 runs in the first over and reached 21 after the second over, showing that the SQL window function correctly tracks the running innings total.

### Query 25: Season-Wise Team Win Ranking

#### Business Problem

Calculate the matches played, matches won and win percentage of every team in each season, and rank teams based on their total wins.

In [48]:
con.execute("""
WITH team_matches AS (

    SELECT
        season,
        team1 AS team
    FROM matches_clean

    UNION ALL

    SELECT
        season,
        team2 AS team
    FROM matches_clean
),

matches_played AS (

    SELECT
        season,
        team,
        COUNT(*) AS matches_played

    FROM team_matches

    GROUP BY
        season,
        team
),

matches_won AS (

    SELECT
        season,
        winner AS team,
        COUNT(*) AS matches_won

    FROM matches_clean

    WHERE winner IS NOT NULL

    GROUP BY
        season,
        winner
),

team_performance AS (

    SELECT
        mp.season,
        mp.team,
        mp.matches_played,
        COALESCE(mw.matches_won, 0) AS matches_won,

        ROUND(
            COALESCE(mw.matches_won, 0) * 100.0
            / NULLIF(mp.matches_played, 0),
            2
        ) AS win_percentage

    FROM matches_played mp

    LEFT JOIN matches_won mw
        ON mp.season = mw.season
       AND mp.team = mw.team
),

ranked_teams AS (

    SELECT
        season,
        team,
        matches_played,
        matches_won,
        win_percentage,

        DENSE_RANK() OVER (
            PARTITION BY season
            ORDER BY
                matches_won DESC,
                win_percentage DESC
        ) AS season_rank

    FROM team_performance
)

SELECT
    season,
    season_rank,
    team,
    CAST(matches_played AS INTEGER) AS matches_played,
    CAST(matches_won AS INTEGER) AS matches_won,
    win_percentage

FROM ranked_teams

ORDER BY
    season,
    season_rank,
    team;
""").df()

,season,season_rank,team,matches_played,matches_won,win_percentage
0,2008,1,Rajasthan Royals,16,13,81.25
1,2008,2,Punjab Kings,15,10,66.67
2,2008,3,Chennai Super Kings,16,9,56.25
3,2008,4,Delhi Capitals,14,7,50.00
4,2008,4,Mumbai Indians,14,7,50.00
...,...,...,...,...,...,...
141,2024,4,Lucknow Super Giants,14,7,50.00
142,2024,5,Royal Challengers Bengaluru,15,7,46.67
143,2024,6,Gujarat Titans,12,5,41.67
144,2024,7,Punjab Kings,14,5,35.71


### Observation

Rajasthan Royals were the strongest team in the 2008 season, winning 13 of 16 matches with an 81.25% win rate. The use of DENSE_RANK() preserves tied positions when multiple teams record the same number of wins in a season.